doc/design.md

```diff
@@ -0,0 +1,136 @@
+
+First, the incompatibilities. The Heroku Metrics infrastructure makes
+use of a library called HDR Histogram, which can record the
+distribution of a large number of values at a relatively low cost. The
+cost is typically less than 1KiB in memory, but the serialization format
+is not standardized making it difficult for other agents needing to 
+report to the Heroku Application Metrics endpoints difficult.
+
+An alternative strategy would involve reporting multiple, related
+metrics, namely, sum of squares, sum, min, max, and count for each
+timer metric.  This would allow all of the reporting dynos' values to
+be fairly represented, and we could even derive standard deviation,
+and your typical average latency. However, average latency, and even
+standard deviation of latency isn't all that helpful. What's helpful
+are the tail latencies (p95, p99, p99.9), and often times the median
+latency. These _aren't_ derivable without knowing what the distribution
+looks like, which as we've previously stated, is not representable in
+an easily standardizable way.
+
+## Reporting Metrics to Heroku
+
+When started, the agentmon program expects a URL passed as an argument
+which speaks the protocol outlined below.
+
+While it'd certainly be possible to report metrics to other services,
+and via other means, the only reporter shipping with agentmon is a
+simple reporter for Heroku's Application Metrics service. This 
+endpoint is POST'ed to over HTTPS, with a few optional headers,
+and a simple JSON body.
+
+The `Content-Type` header should be `application/json`. Optionally,
+the header `Measurements-Time` can be set to an RFC3339 timestamp,
+no more than 4 minutes in the past, representing the time the 
+observations are being reported. The 4 minutes is an implementation
+detail of the Heroku Metrics Aggregation system--metrics timestamped
+older than that will not be aggregated.
+
+Additionally, it's good practice to include the number of measurements
+the payload encodes using the `Measurements-Count` header.
+
+The JSON payload is quite simple:
+
+```json
+{
+   "counters": { 
+      COUNTER_NAME_1: COUNTER_VALUE_1,
+      COUNTER_NAME_2: COUNTER_VALUE_2,
+      ...
+      COUNTER_NAME_N: COUNTER_VALUE_N
+   },
+   "gauges": { 
+      GAUGE_NAME_1: GAUGE_VALUE_1,
+      GAUGE_NAME_2: GAUGE_VALUE_2,
+      ...
+      GAUGE_NAME_N: GAUGE_VALUE_N
+   }
+}
+```
+
+The tricky part is that the names must follow statsd conventions,
+which is to say alpha numeric tokens separated by `-`, `_`, and `.`.
+
+An HTTP 400 Bad Request is returned for any request related issues,
+such as invalid JSON body, stale metrics, or missing `Content-Type`.
+A HTTP 200 OK, with no body is returned on success.
+
+## Receiving Metrics via statsd over UDP
+
+When the program is started with `-statsd-addr IPV4:PORT`, the program
+creates a UDP listener to receive UDP packets, with statsd formatted
+measurements contained with them. Statsd style counts and gauges will
```

Comment:
the program creates a UDP listener to receive UDP packets containing statsd formatted measurements.